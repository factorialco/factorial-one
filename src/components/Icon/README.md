# Icon Component

The Icon component is used to render SVG icons from the F0 Design System in React Native applications.

## Import Pattern

```tsx
// Import the Icon component
import { Icon } from "@factorialco/f0-react-native";

// Import specific icons from app or modules directories
import { AppIcons, ModuleIcons } from "@factorialco/f0-react-native";

// Use the Icon component with the icon as a prop
<Icon icon={AppIcons.Archive} size="md" />
<Icon icon={ModuleIcons.Home} size="lg" />
```

## Props

| Prop      | Type                         | Default | Description                             |
| --------- | ---------------------------- | ------- | --------------------------------------- |
| icon      | IconType                     |         | The icon component to render            |
| size      | "xs" \| "sm" \| "md" \| "lg" | "md"    | The size of the icon                    |
| className | string                       |         | Additional classes for styling the icon |
| testID    | string                       |         | Test ID for testing                     |
| ...props  | SvgProps                     |         | Additional props for the SVG component  |

## Styling with NativeWind

The Icon component is designed to work with NativeWind, allowing you to style icons using Tailwind CSS classes:

```tsx
// Style using Tailwind CSS classes
<Icon icon={AppIcons.Archive} className="text-f1-icon-secondary" />
```

## Available Icons

The library includes two sets of icons:

1. **App Icons** - General purpose icons used throughout the application
2. **Module Icons** - Icons representing specific modules in the Factorial application

## Examples

```tsx
import { Icon } from "@factorialco/f0-react-native";
import { AppIcons, ModuleIcons } from "@factorialco/f0-react-native";

// Basic usage
<Icon icon={AppIcons.Calendar} />

// With size variant
<Icon icon={AppIcons.ChevronDown} size="xs" />
<Icon icon={AppIcons.Check} size="sm" />
<Icon icon={ModuleIcons.Home} size="md" />
<Icon icon={ModuleIcons.Settings} size="lg" />

// With color styling
<Icon icon={AppIcons.Heart} className="text-red-500" />
<Icon icon={AppIcons.InfoCircle} className="text-blue-500" />
```
