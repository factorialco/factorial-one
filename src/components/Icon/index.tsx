import { cva, type VariantProps } from "cva";
import React, {
  forwardRef,
  ForwardRefExoticComponent,
  RefAttributes,
} from "react";
import { SvgProps, Svg } from "react-native-svg";
import { cn } from "../../lib/utils";
import { cssInterop } from "nativewind";

const iconVariants = cva({
  base: "shrink-0",
  variants: {
    size: {
      xl: "w-8 h-8 stroke-xl",
      lg: "w-6 h-6 stroke-lg",
      md: "w-5 h-5 stroke-md",
      sm: "w-4 h-4 stroke-sm",
      xs: "w-3 h-3 stroke-xs",
    },
  },
  defaultVariants: {
    size: "md",
  },
});

export interface IconProps extends SvgProps, VariantProps<typeof iconVariants> {
  icon: IconType;
  testID?: string;
  className?: string;
  variant?:
    | "default"
    | "critical"
    | "neutral"
    | "ghost"
    | "outline"
    | "promote";
  isPressed?: boolean;
}

export type IconType = ForwardRefExoticComponent<
  SvgProps &
    RefAttributes<Svg> & {
      className?: string;
    }
>;

// Keep track of icons that have already had cssInterop applied
const interopAppliedIcons = new WeakSet();

// Function to apply NativeWind interop to an icon component
export function applyIconInterop(icon: IconType) {
  if (!interopAppliedIcons.has(icon)) {
    cssInterop(icon, {
      className: {
        target: "style",
        nativeStyleToProp: {
          color: true,
          opacity: true,
        },
      },
    });
    interopAppliedIcons.add(icon);
  }
  return icon;
}

export const Icon = forwardRef<Svg, IconProps>(function Icon(
  { size, icon, className, testID, ...props },
  ref,
) {
  if (!icon) return null;

  // Apply NativeWind interop to the icon if not already applied
  const Component = applyIconInterop(icon);

  return (
    <Component
      ref={ref}
      {...props}
      className={cn(iconVariants({ size }), className)}
      testID={testID}
    />
  );
});
