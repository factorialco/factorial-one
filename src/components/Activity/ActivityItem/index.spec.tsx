import { render, fireEvent, screen } from "@testing-library/react-native";
import React from "react";
import { IconType } from "../../Icon";
import {
  ActivityItem,
  ActivityItemSkeleton,
} from "../../Activity/ActivityItem";

// Mock the IconAvatar component
jest.mock("../../Avatars/IconAvatar", () => ({
  IconAvatar: () => null,
}));

// Mock dependencies
const mockIcon: IconType = "check" as unknown as IconType;
const mockOnPress = jest.fn();

describe("ActivityItem", () => {
  const defaultProps = {
    id: "activity-123",
    date: "Today",
    title: "Activity Title",
    category: "Time off",
    onPress: mockOnPress,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Snapshot", () => {
    const { toJSON } = render(<ActivityItem {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with required props", () => {
    render(<ActivityItem {...defaultProps} />);

    const title = screen.getByText("Activity Title");
    const moduleDate = screen.getByText("Time off Â· Today");

    expect(title).toBeDefined();
    expect(moduleDate).toBeDefined();
  });

  it("renders correctly with description", () => {
    render(<ActivityItem {...defaultProps} description="Test description" />);

    const description = screen.getByText("Test description");

    expect(description).toBeDefined();
  });

  it("renders icon container when icon is provided", () => {
    render(<ActivityItem {...defaultProps} icon={mockIcon} />);

    const icon = screen.getByLabelText("activity-icon-container");

    expect(icon).toBeDefined();
  });

  it("does not render icon container when icon is not provided", () => {
    render(<ActivityItem {...defaultProps} />);

    const icon = screen.queryByLabelText("activity-icon-container");

    expect(icon).toBeNull();
  });

  it("shows unread indicator when isUnread is true", () => {
    render(<ActivityItem {...defaultProps} isUnread={true} />);

    const indicator = screen.getByLabelText("unread-indicator");

    expect(indicator).toBeDefined();
  });

  it("does not show unread indicator when isUnread is false", () => {
    render(<ActivityItem {...defaultProps} isUnread={false} />);

    const indicator = screen.queryByLabelText("unread-indicator");

    expect(indicator).toBeNull();
  });

  it("calls onPress with correct id when pressed", () => {
    render(<ActivityItem {...defaultProps} />);

    fireEvent.press(screen.getByLabelText("activity-item"));

    expect(mockOnPress).toHaveBeenCalledWith("activity-123");
  });
});

describe("ActivityItemSkeleton", () => {
  it("renders skeleton correctly", () => {
    render(<ActivityItemSkeleton />);

    const skeleton = screen.getByLabelText("activity-skeleton");

    expect(skeleton).toBeDefined();
  });
});
