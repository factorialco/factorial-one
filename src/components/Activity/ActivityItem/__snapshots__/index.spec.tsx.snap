// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityItem Snapshot 1`] = `
<View
  accessibilityLabel="activity-item"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="flex w-full flex-row gap-2 rounded-lg p-2 pr-3 "
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    className="flex-1"
  >
    <Text
      className="line-clamp-1 text-lg font-medium text-f1-foreground"
    >
      Activity Title
    </Text>
    <Text
      className="line-clamp-2 text-lg text-f1-foreground-secondary"
    />
    <View
      className="mt-1.5 flex flex-row"
    >
      <Text
        className="text-md text-f1-foreground-secondary"
      >
        Time off Â· Today
      </Text>
    </View>
  </View>
  <View
    className="ml-1"
  />
</View>
`;
